File: src/lib/canvas.ts
import { deleteCanvasDoc, listCanvases, loadCanvasDoc, saveCanvasDoc } from './bridge'
import {
  CANVAS_VERSION,
  deserializeCanvas,
  serializeCanvas,
  strandedPersonas,
  type CanvasMeta,
} from './persist'
import { personaFromNode } from './library'
import { DEFAULT_ORCHESTRA } from './types'
import { justWentQuiet, playDone } from './chime'
import { useStore } from './store'

/** Which canvas to reopen on launch. Per-machine, so not part of the file. */
const LAST_KEY = 'canvastrator.lastCanvas'

const rid = () => Math.random().toString(36).slice(2, 10)

/** Long enough that a streaming turn doesn't write a file per token. */
const AUTOSAVE_MS = 1200

/**
 * Set while we're the ones replacing the graph. Loading a canvas changes every
 * node, which would otherwise look like an edit and mark the canvas dirty.
 */
let loading = false

let timer: ReturnType<typeof setTimeout> | null = null

/**
 * Bumped whenever the canvas on screen is replaced — opened, cleared, deleted.
 *
 * A save is not instant, and what it does *after* the write is the dangerous
 * half: it records the id as the last canvas and writes it back into the
 * store. Do that for a canvas that was deleted while the write was in flight
 * and you have resurrected it — the pointer says to reopen it, and the next
 * launch does. Catching the id afterwards is not enough either, because a new
 * canvas can have been opened in the meantime and would be overwritten by the
 * finishing save. So the save takes a ticket, and hands nothing back if the
 * canvas moved on while it was away.
 */
let epoch = 0

/**
 * The write currently in flight, so a delete can wait for it.
 *
 * Ordering, not exclusion: without this, a save that started before the delete
 * lands after it, and writes the file straight back out of the snapshot it was
 * already holding. The delete then has nothing left to remove — it already ran.
 */
let pendingWrite: Promise<unknown> | null = null

/** Canvas id we've already warned about, so the log isn't spammed per edit. */
let emptyGuardTripped: string | null = null

/**
 * What the Rust side says when the file is simply gone. Anything else is a
 * canvas of the user's that failed to open, which is a different situation.
 */
const MISSING = 'canvas-missing'

/**
 * The window between the app starting and the launch restore settling.
 *
 * Autosave is watching before the restore has landed, and for that moment the
 * store looks exactly like a brand-new canvas: no id, and any node that
 * appears would mint one. That is one of the two ways the app used to grow an
 * "untitled" on every launch — the file it minted was orphaned a moment later
 * when the real canvas loaded over it, and stayed in the list for ever.
 */
let launchPending = false

/**
 * Set when the launch restore failed for a reason other than the file being
 * gone — an unreadable canvas, a bad write, a disk that did not answer.
 *
 * The other way an "untitled" appeared: the failure was swallowed, the pointer
 * to the canvas was thrown away, and the next node minted a fresh file. So the
 * user's canvas was still on disk, and their work carried on in a new one,
 * silently, every single launch. While this is set nothing may mint a new
 * canvas — the error is on screen, and starting a new one is a decision for
 * the user to make rather than a side effect of typing.
 */
let detached = false

/** Called synchronously at startup, before anything can be added to the graph. */
export function beginLaunchRestore() {
  launchPending = true
}

function suppress<T>(fn: () => T): T {
  loading = true
  try {
    return fn()
  } finally {
    // Cleared after the subscribers for this update have run.
    queueMicrotask(() => {
      loading = false
    })
  }
}

function rememberLast(id: string | null) {
  try {
    if (id) localStorage.setItem(LAST_KEY, id)
    else localStorage.removeItem(LAST_KEY)
  } catch {
    /* private mode, or no storage — reopening is a convenience, not a feature */
  }
}

/**
 * Reopen whatever was on screen last time. A canvas deleted since then just
 * leaves an empty surface, which is the same as a first run.
 */
export async function restoreLastCanvas(): Promise<boolean> {
  try {
    let last: string | null = null
    try {
      last = localStorage.getItem(LAST_KEY)
    } catch {
      return false
    }
    if (!last) return false

    const ok = await openCanvas(last)
    if (ok) return true

    const why = useStore.getState().canvasError ?? ''
    if (why.includes(MISSING)) {
      // Deleted since last time. Nothing to say — an empty surface is the
      // same as a first run, and the pointer is dead weight.
      rememberLast(null)
      useStore.setState({ canvasError: null })
      return false
    }

    // It exists and would not open. Keep pointing at it: the file is the
    // user's work, and the next launch should try again rather than having
    // quietly moved on. Nothing autosaves until they choose what to do.
    detached = true
    useStore.setState({
      canvasError: `Couldn't open your last canvas — ${why}. Pick one from the sidebar, or start a new canvas. Nothing is being saved until you do.`,
    })
    return false
  } finally {
    launchPending = false
  }
}

export const listSavedCanvases = (): Promise<CanvasMeta[]> => listCanvases()

/**
 * Write the current graph to disk. Without a name the canvas keeps the one it
 * has; the first save of an unnamed canvas becomes "untitled".
 */
export async function saveCanvas(name?: string): Promise<boolean> {
  const s = useStore.getState()
  const id = s.canvasId ?? `canvas_${rid()}`
  const canvasName = (name ?? s.canvasName).trim() || 'untitled'
  const updatedAt = Date.now()

  // Snapshot first: a streaming turn can land more text while the write is in
  // flight, and that edit must not be marked as saved.
  const written = useStore.getState()
  const mine = epoch
  try {
    const write = saveCanvasDoc({
      id,
      name: canvasName,
      updatedAt,
      version: CANVAS_VERSION,
      data: serializeCanvas(written),
    })
    pendingWrite = write
    try {
      await write
    } finally {
      if (pendingWrite === write) pendingWrite = null
    }
  } catch (e) {
    // A canvas that was deleted mid-write fails here on some paths; that is
    // the delete working, not an error to put in front of the user.
    if (mine === epoch) useStore.setState({ canvasError: String(e) })
    return false
  }

  // The canvas moved on while this was in flight — deleted, cleared, or
  // another one opened. The bytes are written and there is nothing to be done
  // about that, but this must not point the app back at it.
  if (mine !== epoch) return false

  const now = useStore.getState()
  rememberLast(id)
  useStore.setState({
    canvasId: id,
    canvasName,
    canvasSavedAt: updatedAt,
    canvasDirty:
      now.nodes !== written.nodes ||
      now.edges !== written.edges ||
      now.bus !== written.bus ||
      now.globalRules !== written.globalRules ||
      now.plan !== written.plan,
    canvasError: null,
  })
  return true
}

/**
 * Rename in place. A saved canvas is written straight away so the new name
 * survives a crash; an unsaved one just carries the name to its first save.
 */
export async function renameCanvas(name: string): Promise<boolean> {
  const trimmed = name.trim()
  if (!trimmed) return false
  const s = useStore.getState()
  if (trimmed === s.canvasName) return true

  useStore.setState({ canvasName: trimmed })
  if (!s.canvasId) return true
  return saveCanvas(trimmed)
}

/** Save under a fresh id, leaving the canvas it was forked from untouched. */
export async function saveCanvasAs(name: string): Promise<boolean> {
  useStore.setState({ canvasId: null })
  return saveCanvas(name)
}

export async function openCanvas(id: string): Promise<boolean> {
  let doc
  try {
    doc = await loadCanvasDoc(id)
  } catch (e) {
    useStore.setState({ canvasError: String(e) })
    return false
  }

  const restored = deserializeCanvas(doc.data)
  rescuePersonas(doc.data)
  rememberLast(doc.id)
  detached = false
  epoch++
  suppress(() =>
    useStore.setState({
      ...restored,
      // A saved canvas from another machine may carry no cwd; keep ours.
      cwd: restored.cwd || useStore.getState().cwd,
      canvasId: doc.id,
      canvasName: doc.name,
      canvasSavedAt: doc.updatedAt,
      canvasDirty: false,
      canvasError: null,
      canvasDialog: null,
      selectedId: null,
      openFilePath: null,
    }),
  )
  return true
}

/**
 * Move any persona defined on an old canvas into the library before the node
 * carrying it is dropped. Matched by name, so opening the same canvas twice
 * doesn't duplicate anything, and a persona the user has since edited in the
 * library wins over the stale copy on the canvas.
 */
function rescuePersonas(data: Parameters<typeof strandedPersonas>[0]) {
  const stranded = strandedPersonas(data)
  if (!stranded.length) return

  const library = useStore.getState().library
  const known = new Set(library.map((p) => p.name.trim().toLowerCase()))
  const fresh = stranded
    .filter((d) => !known.has(d.name.trim().toLowerCase()))
    .filter((d, i, all) => all.findIndex((x) => x.name === d.name) === i)
    .map(personaFromNode)
  if (fresh.length) void useStore.getState().setLibrary([...library, ...fresh])
}

/** Start over. The canvas on disk, if any, is left alone. */
export function newCanvas() {
  if (timer) clearTimeout(timer)
  timer = null
  epoch++
  rememberLast(null)
  // An explicit fresh start is exactly the decision the detached state was
  // waiting for, so autosave may mint again.
  detached = false
  suppress(() =>
    useStore.setState({
      nodes: [],
      edges: [],
      bus: [],
      delivered: {},
      selectedId: null,
      openFilePath: null,
      canvasId: null,
      canvasName: 'untitled',
      canvasSavedAt: null,
      canvasDirty: false,
      canvasError: null,
      canvasDialog: null,
      globalRules: '',
      plan: null,
      orchestra: { ...DEFAULT_ORCHESTRA },
    }),
  )
}

/**
 * Everything that has to stop before a file can be removed.
 *
 * A delete competes with autosave twice over: a debounced save may be seconds
 * from firing, and one may already be in flight holding a snapshot of the
 * canvas about to go. Both write the file back — so the row returns, or worse,
 * the file is gone from the list and back on disk, and the next launch reopens
 * the canvas the user deleted.
 */
async function quiesce() {
  if (timer) clearTimeout(timer)
  timer = null
  // A failed write is still a finished one, and that is all this waits for.
  await pendingWrite?.catch(() => {})
}

/** Delete a saved canvas. */
export async function deleteCanvas(id: string): Promise<boolean> {
  await quiesce()
  // Before the delete, so a save that slips in behind it knows the canvas has
  // moved on and keeps its hands off the pointer.
  if (useStore.getState().canvasId === id) epoch++

  try {
    await deleteCanvasDoc(id)
  } catch (e) {
    useStore.setState({ canvasError: String(e) })
    return false
  }
  forgetIfOnScreen(id)
  return true
}

/**
 * Delete several at once, for clearing out canvases that have piled up.
 * One unreadable file shouldn't strand the rest, so a failure is recorded and
 * the sweep carries on; the caller gets the ids that actually went.
 */
export async function deleteCanvases(ids: string[]): Promise<string[]> {
  await quiesce()
  if (ids.includes(useStore.getState().canvasId ?? '')) epoch++

  const gone: string[] = []
  const failed: string[] = []

  for (const id of ids) {
    try {
      await deleteCanvasDoc(id)
      gone.push(id)
      forgetIfOnScreen(id)
    } catch {
      failed.push(id)
    }
  }

  useStore.setState({
    canvasError: failed.length ? `could not delete ${failed.length} of ${ids.length}` : null,
  })
  return gone
}

/**
 * Deleting the canvas that's on screen leaves you on a fresh blank one.
 *
 * It used to just drop the id and keep the graph, on the theory that the work
 * shouldn't die with the file. But a graph with content and no file is exactly
 * what autosave exists to rescue: 1.2s later it minted a new id and wrote the
 * canvas straight back, under the same name. The row reappeared and the delete
 * looked broken. Clearing the board is both what the user asked for and the
 * only way the deletion sticks.
 */
function forgetIfOnScreen(id: string) {
  if (useStore.getState().canvasId !== id) return
  newCanvas()
}

/**
 * Track edits to the graph and autosave them.
 *
 * A canvas with content autosaves even when it has never been saved by hand:
 * it mints an id and lands as "untitled". Requiring an explicit first save
 * meant everything before that save could still be lost, which is exactly what
 * autosave is supposed to prevent. An *empty* canvas still writes nothing —
 * launching the app shouldn't litter the data dir with blank files.
 */
/**
 * Re-run the layout when the *set* of nodes changes — a node appearing or
 * disappearing. Deliberately not on every store change: re-laying out while an
 * agent streams would drag the canvas out from under whoever is reading it.
 * Moving a node by hand doesn't trigger it either, or tidy would fight the user.
 */
/**
 * Chime when the last working agent finishes.
 *
 * Per-agent would be noise on a busy canvas; the useful signal is that the
 * whole canvas has gone quiet and there's nothing left to wait for.
 */
export function watchCompletion(): () => void {
  const busyNow = () =>
    useStore
      .getState()
      .nodes.some(
        (n) => n.type === 'session' && (n.data.state === 'thinking' || n.data.state === 'streaming'),
      )

  let wasBusy = busyNow()

  return useStore.subscribe(() => {
    const isBusy = busyNow()
    if (justWentQuiet(wasBusy, isBusy)) playDone()
    wasBusy = isBusy
  })
}

export function watchLayout(): () => void {
  let known = new Set(useStore.getState().nodes.map((n) => n.id))
  let timer: ReturnType<typeof setTimeout> | null = null

  const unsubscribe = useStore.subscribe((s) => {
    const ids = new Set(s.nodes.map((n) => n.id))
    const changed = ids.size !== known.size || [...ids].some((id) => !known.has(id))
    known = ids
    if (!changed || !s.autoTidy || loading) return

    if (timer) clearTimeout(timer)
    // Long enough that a burst of file nodes settles into one pass.
    timer = setTimeout(() => {
      timer = null
      // Never reposition a node the user is holding.
      if (useStore.getState().nodes.some((n) => n.dragging)) return
      useStore.getState().tidy()
    }, 700)
  })

  return () => {
    if (timer) clearTimeout(timer)
    unsubscribe()
  }
}

export function watchCanvas(): () => void {
  let seen = useStore.getState()

  const unsubscribe = useStore.subscribe((s) => {
    const changed =
      s.nodes !== seen.nodes ||
      s.edges !== seen.edges ||
      s.bus !== seen.bus ||
      s.globalRules !== seen.globalRules ||
      // A plan is work the user agreed to, or is about to; losing it to a
      // crash would mean reading the orchestrator's reasoning over again.
      s.plan !== seen.plan ||
      s.planning !== seen.planning ||
      s.orchestra !== seen.orchestra
    seen = s
    if (!changed || loading) return

    if (!s.canvasDirty) useStore.setState({ canvasDirty: true })
    if (!s.nodes.length) {
      // Never let autosave write an empty graph. Reaching zero nodes is far
      // more often a bug or a mis-click than an edit worth persisting, and the
      // saved canvas is the only copy. An explicit save still goes through.
      if (!s.canvasId) return
      if (emptyGuardTripped !== s.canvasId) {
        emptyGuardTripped = s.canvasId
        console.warn('canvastrator: canvas is empty — autosave skipped to protect the saved copy')
      }
      return
    }
    emptyGuardTripped = null

    // Minting a *new* canvas is the only thing gated here. A canvas that
    // already has a file goes on saving to it either way — the risk being
    // guarded against is a second file, not a lost edit.
    if (!s.canvasId && (launchPending || detached)) return

    if (timer) clearTimeout(timer)
    timer = setTimeout(() => {
      timer = null
      void saveCanvas()
    }, AUTOSAVE_MS)
  })

  return () => {
    if (timer) clearTimeout(timer)
    timer = null
    unsubscribe()
  }
}

File: src-tauri/src/canvas.rs
//! Saved canvases: the whole graph, on disk, under the app's data dir.
//!
//! The frontend owns the shape of `data` — this side only cares about the
//! envelope (id, name, timestamp) so it can list canvases without the UI
//! having to open every file.

use std::path::PathBuf;

use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Manager};

#[derive(Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct CanvasDoc {
    pub id: String,
    pub name: String,
    pub updated_at: i64,
    /// When the canvas was first written. Fixed for the life of the file, so
    /// the list can be ordered by something that saving does not move.
    ///
    /// Optional because canvases written before this field existed do not
    /// carry it; those fall back to `updated_at` until their next save, which
    /// stamps the timestamp they had at that point rather than the clock —
    /// otherwise opening an old canvas would send it to the end of the list.
    #[serde(default)]
    pub created_at: Option<i64>,
    pub version: u32,
    pub data: serde_json::Value,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CanvasMeta {
    pub id: String,
    pub name: String,
    pub updated_at: i64,
    /// Never null here: a file without one is reported as its own timestamp.
    pub created_at: i64,
    pub version: u32,
    /// Cheap "is this the canvas I meant" signal in the open list.
    pub nodes: usize,
}

/// Ids become filenames, so anything that could escape the directory is out.
fn valid_id(id: &str) -> bool {
    !id.is_empty()
        && id.len() <= 64
        && id
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || c == '-' || c == '_')
}

fn dir(app: &AppHandle) -> Result<PathBuf, String> {
    let base = app
        .path()
        .app_data_dir()
        .map_err(|e| format!("no app data dir: {e}"))?
        .join("canvases");
    std::fs::create_dir_all(&base).map_err(|e| format!("{}: {e}", base.display()))?;
    Ok(base)
}

fn path_for(app: &AppHandle, id: &str) -> Result<PathBuf, String> {
    if !valid_id(id) {
        return Err(format!("invalid canvas id: {id}"));
    }
    Ok(dir(app)?.join(format!("{id}.json")))
}

#[tauri::command]
pub fn list_canvases(app: AppHandle) -> Result<Vec<CanvasMeta>, String> {
    let base = dir(&app)?;
    let entries = std::fs::read_dir(&base).map_err(|e| format!("{}: {e}", base.display()))?;

    let mut out: Vec<CanvasMeta> = entries
        .flatten()
        .filter(|e| e.path().extension().is_some_and(|x| x == "json"))
        // A single unreadable or half-written file must not hide the rest.
        .filter_map(|e| std::fs::read_to_string(e.path()).ok())
        .filter_map(|raw| serde_json::from_str::<CanvasDoc>(&raw).ok())
        .map(|doc| CanvasMeta {
            nodes: doc
                .data
                .get("nodes")
                .and_then(|n| n.as_array())
                .map_or(0, |a| a.len()),
            created_at: doc.created_at.unwrap_or(doc.updated_at),
            id: doc.id,
            name: doc.name,
            updated_at: doc.updated_at,
            version: doc.version,
        })
        .collect();

    // Creation order, oldest first. The list used to be sorted by the save
    // timestamp, which meant that opening a canvas — and so autosaving it —
    // pulled its row to the top under the pointer that had just clicked it.
    // A switcher whose rows move when you use them is one you cannot learn,
    // so the order is fixed at creation; the id breaks ties between canvases
    // saved in the same millisecond so it is total either way.
    out.sort_by(|a, b| {
        a.created_at
            .cmp(&b.created_at)
            .then_with(|| a.id.cmp(&b.id))
    });
    Ok(out)
}

/// Returned when the canvas file simply is not there any more.
///
/// The frontend has to tell this apart from every other failure. A canvas the
/// user deleted is nothing to complain about — reopening it is a convenience —
/// while one that is present but unreadable is a canvas of theirs that failed
/// to open, and treating the two the same is how the pointer to it gets thrown
/// away and the work starts again in a new file.
pub const MISSING: &str = "canvas-missing";

#[tauri::command]
pub fn load_canvas(app: AppHandle, id: String) -> Result<CanvasDoc, String> {
    let path = path_for(&app, &id)?;
    if !path.exists() {
        return Err(MISSING.to_string());
    }
    let raw = std::fs::read_to_string(&path).map_err(|e| format!("{id}: {e}"))?;
    serde_json::from_str(&raw).map_err(|e| format!("{id} is not a canvas file: {e}"))
}

#[tauri::command]
pub fn save_canvas(app: AppHandle, mut doc: CanvasDoc) -> Result<(), String> {
    let path = path_for(&app, &doc.id)?;
    // The creation stamp belongs to the file, not to whoever is writing it:
    // resolving it here means no caller can drop it and reorder the list by
    // accident.
    doc.created_at = Some(resolve_created_at(&doc, read_doc(&path).as_ref()));
    let body = serde_json::to_string(&doc).map_err(|e| e.to_string())?;
    // Write-then-rename: a crash mid-save leaves the previous canvas intact
    // rather than a truncated file that no longer parses.
    let tmp = path.with_extension("json.tmp");
    std::fs::write(&tmp, body).map_err(|e| format!("{}: {e}", tmp.display()))?;
    std::fs::rename(&tmp, &path).map_err(|e| format!("{}: {e}", path.display()))
}

fn read_doc(path: &std::path::Path) -> Option<CanvasDoc> {
    let raw = std::fs::read_to_string(path).ok()?;
    serde_json::from_str(&raw).ok()
}

/// The creation stamp a save should carry.
///
/// Whatever the incoming doc already claims wins, then whatever is on disk —
/// including an older file's save timestamp, which is the closest thing to a
/// creation time it has. Only a canvas nobody has written before is stamped
/// with the clock it arrived with.
fn resolve_created_at(doc: &CanvasDoc, existing: Option<&CanvasDoc>) -> i64 {
    doc.created_at
        .or_else(|| existing.and_then(|e| e.created_at))
        .or_else(|| existing.map(|e| e.updated_at))
        .unwrap_or(doc.updated_at)
}

#[tauri::command]
pub fn delete_canvas(app: AppHandle, id: String) -> Result<(), String> {
    let path = path_for(&app, &id)?;
    match std::fs::remove_file(&path) {
        Ok(()) => Ok(()),
        // Already gone is the outcome the caller wanted.
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(()),
        Err(e) => Err(format!("{id}: {e}")),
    }
}

#[cfg(test)]
mod tests {
    use super::{resolve_created_at, valid_id, CanvasDoc};

    fn doc(created_at: Option<i64>, updated_at: i64) -> CanvasDoc {
        CanvasDoc {
            id: "c1".into(),
            name: "c".into(),
            updated_at,
            created_at,
            version: 1,
            data: serde_json::Value::Null,
        }
    }

    #[test]
    fn saving_does_not_move_a_canvas_creation_stamp() {
        // A canvas that has been saved before keeps the stamp on disk, however
        // many times it is written afterwards.
        let on_disk = doc(Some(100), 100);
        assert_eq!(resolve_created_at(&doc(None, 900), Some(&on_disk)), 100);
        // One written before the field existed adopts its last save time, not
        // the clock — opening an old canvas must not send it to the end.
        let legacy = doc(None, 250);
        assert_eq!(resolve_created_at(&doc(None, 900), Some(&legacy)), 250);
        // Only a genuinely new file takes the incoming timestamp.
        assert_eq!(resolve_created_at(&doc(None, 900), None), 900);
    }

    #[test]
    fn ids_that_could_escape_the_directory_are_rejected() {
        assert!(valid_id("canvas_ab12"));
        assert!(valid_id("a-b_C9"));
        assert!(!valid_id(""));
        assert!(!valid_id("../etc/passwd"));
        assert!(!valid_id("a/b"));
        assert!(!valid_id("a.json"));
        assert!(!valid_id(&"x".repeat(65)));
    }
}

/// Where handover notes are kept.
///
/// Beside the canvases in the app's own data directory, deliberately not in the
/// user's repository: a note is written whenever a window fills, which on a
/// long run is several times an hour, and a tool that quietly starts committing
/// files to somebody's project has overstepped. The note carries its working
/// directory in its header, so it can always be read back to where it belongs.
fn notes_dir(app: &AppHandle) -> Result<PathBuf, String> {
    let base = app
        .path()
        .app_data_dir()
        .map_err(|e| format!("no app data dir: {e}"))?
        .join("notes");
    std::fs::create_dir_all(&base).map_err(|e| format!("{}: {e}", base.display()))?;
    Ok(base)
}

/// Write a handover note and return the path it landed at.
#[tauri::command]
pub fn write_note(app: AppHandle, name: String, text: String) -> Result<String, String> {
    // The name is generated, but it ends up as a path: a `..` or a slash that
    // slipped through would write outside the notes directory.
    let clean = name
        .chars()
        .filter(|c| c.is_ascii_alphanumeric() || matches!(c, '-' | '_' | '.'))
        .collect::<String>();
    if clean.is_empty() || clean.starts_with('.') {
        return Err(format!("unusable note name: {name}"));
    }
    let path = notes_dir(&app)?.join(&clean);
    std::fs::write(&path, text).map_err(|e| format!("{}: {e}", path.display()))?;
    Ok(path.to_string_lossy().into_owned())
}
